Take-off throttle slew time

lthall opened the pull request:
This PR adds control over the throttle slew time as the aircraft moves from landed to flying.
The autopilot looks for vertical acceleration, velocity or position changes to detect take-off. If 90% throttle is reached take-off is declared, limiting the maximum time delay.

The primary benefit of this PR is making the take-off behaviour independent of the vertical position controller tune. It will also improve the consistency of take-off when different payloads are used.

lthall commented on ArduCopter/land_detector.cpp:
```diff
@@ -49,12 +49,9 @@ void Copter::update_land_detector()
 #if FRAME_CONFIG == HELI_FRAME
         // if rotor speed and collective pitch are high then clear landing flag
         if (motors->get_takeoff_collective() && motors->get_spool_state() == AP_Motors::SpoolState::THROTTLE_UNLIMITED) {
-#else
```
We need to do the same thing for Heli.

bnsgeyer replied:
I would have to add an additional condition here that would add non manual throttle for this statement because I don't clear the land complete flag in acro or stabilize.  It is all done in this method.

lthall commented:
resolves #6724

chobitsfan commented:
Hi @lthall This is great. I think maybe I could help testing it?

chobitsfan commented:
test flight on cubeblack, f450 frame
[2022-07-11 11-21-20.zip](https://github.com/ArduPilot/ardupilot/files/9080704/2022-07-11.11-21-20.zip)

lthall commented:
> test flight on cubeblack, f450 frame [2022-07-11 11-21-20.zip](https://github.com/ArduPilot/ardupilot/files/9080704/2022-07-11.11-21-20.zip)

Thanks for testing, what did you think?

Looking at your log it looks like it was triggering of something early. You have a pretty low vertical acceleration of 1m/s so the ground noise may be enough to trigger it because it only needs to get to 0.5 m/s/s.
The trigger thresholds are 50% max acceleration, your target climb rate (limited to 10% to 50% max velocity).

I may need to add a counter for consecutive detections to de-bounce the detection......

Thanks a bunch for testing. This shows a nice weakness to fix!!

bnsgeyer commented on ArduCopter/land_detector.cpp:
```diff
@@ -49,12 +49,9 @@ void Copter::update_land_detector()
 #if FRAME_CONFIG == HELI_FRAME
         // if rotor speed and collective pitch are high then clear landing flag
         if (motors->get_takeoff_collective() && motors->get_spool_state() == AP_Motors::SpoolState::THROTTLE_UNLIMITED) {
```
```suggestion
        if (flightmode->has_manual_throttle() && motors->get_takeoff_collective() && motors->get_spool_state() == AP_Motors::SpoolState::THROTTLE_UNLIMITED) {
```

chobitsfan commented:
> @chobitsfan Any chance you could test again once the tests pass?

Sure. I will test it today (or tomorrow)

chobitsfan commented:
I have tested it with updated code. It looks good. Thank you @lthall 

[2022-07-12 13-51-30.zip](https://github.com/ArduPilot/ardupilot/files/9089959/2022-07-12.13-51-30.zip)

chobitsfan commented:
> For some reason the first two flights trigger immediately.

Hi @lthall I am very sorry for late reply. Yes. it seems a little strange. I will try to test it more. Thank you

lthall commented:
@chobitsfan No problems. I need to understand how this can happen and add some logging or something before you test again.

lthall commented:
I have made a little progress. It is only happening in Loiter so it is something in the pilot takeoff.

lthall commented:
@chobitsfan I have added some logging to try to work out why your aircraft is detecting take-off immediately. Could you test with this code please.

lthall commented:
@chobitsfan Ok, I have found and fixed the problem. Are you happy to test again?

chobitsfan commented:
Hi @lthall
 It works fine. Except for once, copter flip over during takeoff in loiter mode. I am unable to produce it again. It probably due to soft grass field. 
[2022-07-21 13-53-47.zip](https://github.com/ArduPilot/ardupilot/files/9156661/2022-07-21.13-53-47.zip)

lthall commented:
> Hi @lthall It works fine. Except for once, copter flip over during takeoff in loiter mode. I am unable to produce it again. It probably due to soft grass field. [2022-07-21 13-53-47.zip](https://github.com/ArduPilot/ardupilot/files/9156661/2022-07-21.13-53-47.zip)

Confirmed the leg got stuck in the ground. The aircraft was doing all it could to pull that side up.

Ok, I am going to remove the counter change and squash this all together. I think we have pretty much got this sorted now. I really appreciate your help @chobitsfan !!!!

lthall commented:
Ok, I have rebased on master, cleaned everything up and pushed again. @chobitsfan if you have some time you could do one more check to ensure we are still good to go once all the checks have passed. I think I am good to get this in then.

chobitsfan commented:
Hi @lthall I just tested rebased code. It works good. Thank you :)
[2022-07-25 09-50-31.zip](https://github.com/ArduPilot/ardupilot/files/9177325/2022-07-25.09-50-31.zip)

tridge commented:
@rmackay9 I'd like to get your perspective on the parameter name. To me TKOFF_THRTIME would be the time in seconds to start takeoff, but this isn't, as you'd need to multiply by the hover throttle.
we could do TKOFF_SLEWRATE, but to be consistent with plane that would need to be a percentage, but @lthall doesn't want any percentages

tridge commented on ArduCopter/Parameters.cpp:
```diff
@@ -1143,6 +1143,13 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Description: This is the time in seconds to ramp the throttle from minimum to maximum during takeoff.
+    // @Range: 0.25 5.0
+    // @User: Standard
+    AP_GROUPINFO("TKOFF_THRTIME", 57, ParametersG2, takeoff_throttle_ramp_time, 2.0),
```
i think the interaction with hover throttle in terms of the actual time to get off the ground is problematic from an end user point of view

lthall replied:
The user currently doesn't know how long it takes to get off the ground so that value is not relevant to the user. 
The actual time to takeoff currently changes each time during a manual takeoff based on the vertical velocity commanded by the pilot. 
In all modes the takeoff time depends on the PSC Z tuning and the vertical climb rate setpoint and weight. 
We also don't know what the actual hover throttle is due to change in payload, fuel load, or even inaccurate settings of hover throttle with learn turned off. (I set hover throttle to be the hover throttle at maximum payload with learn turned off as this is what I think the best balance given the various issues)

The goal here is to provide a consent takeoff behaviour that is tolerant to all the factors above and easy to understand and set for the user. The throttle rate of change is the single variable that we can use to define this independent of all other factors.

So we have two options, takeoff slew time or take off slew rate. 

I chose slew time because I could describe this to a user as "This is how long it takes to move the throttle from minimum to maximum while it is waiting for the aircraft to leave the ground. It is also the maximum time before the aircraft will give up trying to detect take-off and assume it is flying".

In reality most people will simply increase or decrease this number by a factor of 2 if they want a faster or slower take-off.

As a general rule I move from time constant to rates based on the nominal magnitude. So for this case we are on the edge there. We are talking throttle rates of 0.25 to 4 and ramp times or time constants or 0.25 to 4.

lthall commented:
The user currently doesn't know how long it takes to get off the ground so that value is not relevant to the user.
The actual time to takeoff currently changes each time during a manual takeoff based on the vertical velocity commanded by the pilot.
In all modes the takeoff time depends on the PSC Z tuning and the vertical climb rate setpoint and weight.
We also don't know what the actual hover throttle is due to change in payload, fuel load, or even inaccurate settings of hover throttle with learn turned off. (I set hover throttle to be the hover throttle at maximum payload with learn turned off as this is what I think the best balance given the various issues)

The goal here is to provide a consent takeoff behaviour that is tolerant to all the factors above and easy to understand and set for the user. The throttle rate of change is the single variable that we can use to define this independent of all other factors.

So we have two options, takeoff slew time or take off slew rate.

I chose slew time because I could describe this to a user as "This is how long it takes to move the throttle from minimum to maximum while it is waiting for the aircraft to leave the ground. It is also the maximum time before the aircraft will give up trying to detect take-off and assume it is flying".

In reality most people will simply increase or decrease this number by a factor of 2 if they want a faster or slower take-off.

As a general rule I move from time constant to rates based on the nominal magnitude. So for this case we are on the edge there. We are talking throttle rates of 0.25 to 4 and ramp times or time constants or 0.25 to 4.

bnsgeyer commented:
I agree with @lthall that slew time makes more sense.  Most users will not be looking to have the aircraft takeoff in a specific time but will notice if it was too aggressive or too slow, and then adjust the parameter accordingly.  I think time to hover throttle is meaningless and most users have no sense of what a good number would be let alone look to set it to some exact fraction of a second.  I can take slew time and think of applying full throttle/collective in that time and get a physical sense of the rate and how it compares to a piloted takeoff.

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -1137,6 +1137,13 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
 
+    // @Param: TKOFF_THRTIME
+    // @DisplayName: Ramp time of throttle during take-off
+    // @Description: This is the time in seconds to ramp the throttle from minimum to maximum while checking for a succsessful takeoff.
```
We can remove the "This is the" part.

rmackay9 replied:
We could add a Units to this description

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -1137,6 +1137,13 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     AP_SUBGROUPINFO(command_model_pilot, "PILOT_Y_", 56, ParametersG2, AC_CommandModel),
 
+    // @Param: TKOFF_THRTIME
+    // @DisplayName: Ramp time of throttle during take-off
```
Let's start this display name with Takeoff.

lthall replied:
Take-off throttle ramp time.

rmackay9 replied:
How about TKOFF_SPOOL_TIME?

timtuxworth replied:
Spool? To align with plane, but make it clear it's different (and using the actual wording from what Leonard had to say above) can I suggest TKOFF_SLEWTIME makes most sense. Then you have TKOFF_SLEWRATE for plane and TKOFF_SLEWTIME otherwise. Then when @tridge is convinced to change plane, the two parameters are right next to each other in the list.

rmackay9 replied:
TKOFF_SLEWTIME is another decent option.  Here are the other similar parameters within Copter:

- MOT_SLEW_DN_TIME
- MOT_SLEW_UP_TIME
- MOT_SPOOL_TIME

rmackay9 replied:
another option is MOT_THST_TIME.

rmackay9 commented on ArduCopter/takeoff.cpp:
```diff
@@ -130,17 +141,32 @@ void Mode::auto_takeoff_run()
-        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), auto_yaw.rate_cds());
+        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), 0.0);
+        return;
+    } else if (copter.ap.land_complete) {
```
No need for the "else" here I think.  The if can start on a new line.

lthall commented:
I have added a commit addressing the "else" feedback. I will make the parameter name and parameter description name changes based on the feedback on the next DevCall and squash everything before we accept this PR.

rmackay9 commented:
Cool.  Looks like it also have conflicts with master so perhaps rebase it on master too.

lthall commented:
TKOFF_SLEW_TIME is the name chosen in the dev call.

rmackay9 commented on ArduCopter/land_detector.cpp:
```diff
@@ -49,12 +49,9 @@ void Copter::update_land_detector()
         if (motors->get_takeoff_collective() && motors->get_spool_state() == AP_Motors::SpoolState::THROTTLE_UNLIMITED) {
-#else
-        // if throttle output is high then clear landing flag
-        if (motors->get_throttle() > get_non_takeoff_throttle()) {
```
This leaves get_non_takeoff_throttle() so that it is only used within set-mode as a check when switching to a manual throttle.  Maybe that's correct or maybe the check should be simplified to remove the helper function.

rmackay9 replied:
Removing this check may increase the risk that a multicopter gets off the ground in modes with manual throttle (including Stabilize) because it means each mode becomes responsible for setting the land_complete flag.  Stabilize mode for example will only set land_complete = false if the motors library's lower_limit is false.  This means we need to be more careful that each motor library sets this flag correctly.

rmackay9 replied:
Maybe we can put back this check with an additional condition that !takeoff.running().  This could close the a potential gap.

lthall replied:
I have done this in addition to explicitly handling leaving takeoff before it is completed.

rmackay9 commented on ArduCopter/takeoff.cpp:
```diff
@@ -130,17 +141,35 @@ void Mode::auto_takeoff_run()
+        pos_control->update_xy_controller();
+        attitude_control->reset_rate_controller_I_terms();
+        attitude_control->input_thrust_vector_rate_heading(pos_control->get_thrust_vector(), 0.0);
+        if (throttle >= 0.9 || 
```
Looks like this check is duplicate code from above.  It's complex enough that it should be moved to a helper function I think.

lthall replied:
This isn't quiet duplicate code. Idealy I would like to combine both takeoff functions into a single function.

rmackay9 commented on ArduCopter/takeoff.cpp:
```diff
@@ -85,16 +80,32 @@ void Mode::_TakeOff::do_pilot_takeoff(float& pilot_climb_rate_cm)
+            (copter.pos_control->get_z_accel_cmss() >= 0.5 * copter.pos_control->get_max_accel_z_cmss()) ||
+            (copter.pos_control->get_vel_desired_cms().z >= constrain_float(pilot_climb_rate_cm, copter.pos_control->get_max_speed_up_cms() * 0.1, copter.pos_control->get_max_speed_up_cms() * 0.5)) || 
+            (is_positive(take_off_complete_alt) && copter.pos_control->get_pos_target_z_cm() - take_off_start_alt > 0.5 * (take_off_complete_alt  - take_off_start_alt))) {
+            copter.set_land_complete(false);
```
We need to remember that the takeoff can be interrupted if the user switches modes during takeoff.  I worry that this could lead to this bit of takeoff detection code not running and leaving the vehicle with land_complete = false.

rmackay9 commented on ArduCopter/mode_throw.cpp:
```diff
@@ -52,6 +52,7 @@ void ModeThrow::run()
 
     } else if (stage == Throw_Detecting && throw_detected()){
         gcs().send_text(MAV_SEVERITY_INFO,"throw detected - spooling motors");
+        copter.set_land_complete(false);
```
My guess is that this is an example where the autotune turned up a situation where the mode was not fulfilling it's new responsibility to set land_complete = false.  Previously it was likely being caught by the overall check in update_land_detector().

I really think we are opening ourselves up to small programming errors in a single mode (which may not be well tested) leading to rare but reputation damaging fall-out-of-the-sky situations.

lthall replied:
No, I new this was here and didn't require auto test. There are only three ways an aircraft leaves the ground. Manual, takeoff (manual and auto) and throw.

rmackay9 commented on ArduCopter/takeoff.cpp:
```diff
@@ -130,17 +146,35 @@ void Mode::auto_takeoff_run()
+    }
+    
+    // aircraft stays in landed state until vertical movement is detected or 90% throttle is reached
+    if (copter.ap.land_complete) {
```
The conditions in these two code blocks, (e.g. "motors->get_spool_state() != AP_Motors::SpoolState::THROTTLE_UNLIMITED" and "copter.ap.land_complete") could theoretically at least both be false meaning that the position controller and attitude controllers would not be updated.

lthall commented:
This PR is rebased and cleaned up and I believe I have addressed all the concerns by adding a backup check that should revert to the existing behaviour if a corner case is found. This also calls a flow control error.

lthall commented:
I have retested this and everything is working as expected.

bnsgeyer commented:
@lthall I think the line of code you pointed me to is fine for heli.  I did some quick testing in real flight simulator.  I thought the not landed flag is set on the actual initial climb rate but it appears that it is being set almost immediately after setting the auto mode.  Is that by design?  Also you can see in the plot that the desired climb rate is not tied to the actual climb rate while the aircraft is on the ground and the collective is being raised.  Then it appears to be reset at the point the aircraft takes off.  Would it not be better to have the desired climb rate just track the actual until lift off?

![image](https://user-images.githubusercontent.com/9907906/187587451-571a730e-fdd0-4edd-9ea7-ae644aedb4a1.png)

lthall commented:
>> I think the line of code you pointed me to is fine for heli.

Thanks Bill.

>> I thought the not landed flag is set on the actual initial climb rate but it appears that it is being set almost immediately after setting the auto mode.

I need to see the log but it looks like the heli detected take-off immediately. It may have been due to the non-zero throttle on entering Auto, or maybe there is some acceleration noise in the simulator that is triggering it immediately. Take-off only resets the climb rate at the end of ramp up so there must be something funny about how this is being graphed or something. I need the log.

Heli simulations look like this for guided:
![image](https://user-images.githubusercontent.com/100896/187602390-72e1d285-0d7e-4021-ab79-f5220cfe5b36.png)

and Auto:
![image](https://user-images.githubusercontent.com/100896/187602784-519aa7d9-b73c-456e-a6d8-b5ae082563c5.png)

lthall commented:
Ok, after a very frustrating day trying to get RealFlight going I have been able to test this myself on heli.

Loiter:
![image](https://user-images.githubusercontent.com/100896/187647339-a5eec715-07d3-4f66-b446-4d92043801e6.png)

Guided:
![image](https://user-images.githubusercontent.com/100896/187647461-87464b0d-36d8-497e-8c85-1b73048ef292.png)

Auto:
![image](https://user-images.githubusercontent.com/100896/187647571-2d27c3b2-68e7-46c4-8d6a-e9a4711c5bc2.png)

The interesting thing is Auto where it looks like the takeoff was triggered early by noise but I don't think we have the update rate to see it.

In any case the worst case is the aircraft takes off like the old code does.

bnsgeyer reviewed (approved):
The changes for Heli are fine. Spoke with Leonard regarding the issue I raised in testing heli. The behavior if it triggers the not landed flag early should be no worse than the current takeoff behavior. I am happy with merging this.
